### ISSUE TITLE
Encountered anon const with inference variable args but no error reported

### ISSUE THREAD

**Issue by nazar-pc** [rust-lang/rust-clippy#13807]
### Summary

This branch: https://github.com/autonomys/subspace/tree/update-rust (commit 0d36b465cf1cf4a40d74d36f3f8a9d6cb56bb924 to be specific)

When running `cargo clippy --all-targets` it fails, while nightly `nightly-2024-10-22` I was upgrading from worked.

### Version

```text
rustc 1.85.0-nightly (a224f3807 2024-12-09)
binary: rustc
commit-hash: a224f3807e58afc9353510f1d556c607d367545d
commit-date: 2024-12-09
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.5
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
error: internal compiler error: Encountered anon const with inference variable args but no error reported
  |
  = note: delayed at compiler/rustc_trait_selection/src/traits/mod.rs:605:27
             0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
             1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
             2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
             3: <rustc_errors::DiagCtxtHandle>::delayed_bug::<&str>
             4: rustc_trait_selection::traits::try_evaluate_const.cold
             5: <rustc_trait_selection::traits::normalize::AssocTypeNormalizer as rustc_type_ir::fold::TypeFolder<rustc_middle::ty::context::TyCtxt>>::fold_const
             6: <rustc_middle::ty::Ty as rustc_type_ir::fold::TypeSuperFoldable<rustc_middle::ty::context::TyCtxt>>::try_super_fold_with::<rustc_trait_selection::traits::normalize::AssocTypeNormalizer>
             7: <rustc_trait_selection::traits::normalize::AssocTypeNormalizer as rustc_type_ir::fold::FallibleTypeFolder<rustc_middle::ty::context::TyCtxt>>::try_fold_binder::<rustc_type_ir::predicate_kind::PredicateKind<rustc_middle::ty::context::TyCtxt>>
             8: <rustc_trait_selection::traits::select::SelectionContext>::confirm_candidate
             9: <rustc_trait_selection::traits::select::SelectionContext>::evaluate_candidate::{closure#0}::{closure#0}
            10: <rustc_trait_selection::traits::select::SelectionContext>::candidate_from_obligation_no_cache
            11: <rustc_trait_selection::traits::select::SelectionContext>::candidate_from_obligation
            12: <rustc_trait_selection::traits::select::SelectionContext>::evaluate_trait_predicate_recursively
            13: <rustc_trait_selection::traits::select::SelectionContext>::evaluate_root_obligation
            14: rustc_traits::evaluate_obligation::evaluate_obligation
            15: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::evaluate_obligation::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 2]>>
            16: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::DefaultCache<rustc_type_ir::canonical::CanonicalQueryInput<rustc_middle::ty::context::TyCtxt, rustc_middle::ty::ParamEnvAnd<rustc_middle::ty::predicate::Predicate>>, rustc_middle::query::erase::Erased<[u8; 2]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
            17: rustc_query_impl::query_impl::evaluate_obligation::get_query_incr::__rust_end_short_backtrace
            18: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt>::evaluate_obligation_no_overflow
            19: <core::iter::adapters::filter::Filter<core::iter::adapters::map::Map<core::slice::iter::Iter<rustc_hir_typeck::method::probe::Candidate>, <rustc_hir_typeck::method::probe::ProbeContext>::consider_candidates::{closure#0}>, <rustc_hir_typeck::method::probe::ProbeContext>::consider_candidates::{closure#1}> as core::iter::traits::iterator::Iterator>::next
            20: <rustc_hir_typeck::method::probe::ProbeContext>::pick_method
            21: <rustc_hir_typeck::method::probe::ProbeContext>::pick_core
            22: <rustc_hir_typeck::fn_ctxt::FnCtxt>::probe_for_name
            23: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_path
            24: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
            25: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
            26: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
            27: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
            28: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
            29: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
            30: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
            31: rustc_hir_typeck::check::check_fn
            32: rustc_hir_typeck::typeck
            33: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 8]>>
            34: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 8]>, rustc_query_system::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
            35: rustc_query_impl::query_impl::typeck::get_query_incr::__rust_end_short_backtrace
            36: <rustc_middle::hir::map::Map>::par_body_owners::<rustc_hir_analysis::check_crate::{closure#4}>::{closure#0}
            37: rustc_hir_analysis::check_crate
            38: rustc_interface::passes::run_required_analyses
            39: rustc_interface::passes::analysis
            40: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
            41: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
            42: rustc_query_impl::query_impl::analysis::get_query_incr::__rust_end_short_backtrace
            43: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
            44: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
            45: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
            46: std::sys::pal::unix::thread::Thread::new::thread_start
            47: start_thread
                       at ./nptl/pthread_create.c:447:8
            48: clone3
                       at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:78

  ```

  </p>
</details>

**Comment by y21** [comment]
No clippy frames in the backtrace, does this also happen with `cargo check`? This looks to me like a bug in rustc's type checking code
